File: python/evaluator/analyzer.py
import sqlite3
from dataclasses import dataclass
from datetime import datetime
from typing import Any, Dict, List, Optional


@dataclass
class TradeRecord:
    ticket: int
    symbol: str
    action: str
    lots: float
    open_price: float
    close_price: Optional[float]
    open_time: str
    close_time: Optional[str]
    profit: float
    comment: Optional[str]


@dataclass
class TradeMetrics:
    total_trades: int
    winning_trades: int
    losing_trades: int
    win_rate: float
    total_profit: float
    gross_profit: float
    gross_loss: float
    profit_factor: float
    max_drawdown: float
    max_drawdown_pct: float
    avg_trade_profit: float
    largest_win: float
    largest_loss: float
    trades: List[TradeRecord]


class TradeAnalyzer:
    def __init__(self, db_path: str = "trade_pipeline.db"):
        self.db_path = db_path

    def load_trades(self, limit: Optional[int] = None) -> List[TradeRecord]:
        try:
            conn = sqlite3.connect(self.db_path)
            cursor = conn.cursor()
            query = """
                SELECT ticket, symbol, action, lots, open_price, close_price, open_time, close_time, profit, comment
                FROM trades
                WHERE close_time IS NOT NULL
                ORDER BY close_time ASC
            """
            if limit:
                query += f" LIMIT {limit}"

            cursor.execute(query)
            rows = cursor.fetchall()
            conn.close()
        except Exception as e:
            # テーブルが存在しないかロックされている場合は空リスト
            return []

        trades = []
        for r in rows:
            trades.append(
                TradeRecord(
                    ticket=r[0],
                    symbol=r[1],
                    action=r[2],
                    lots=r[3],
                    open_price=r[4],
                    close_price=r[5],
                    open_time=r[6],
                    close_time=r[7],
                    profit=r[8],
                    comment=r[9],
                )
            )
        return trades

    def calculate_metrics(self, trades: Optional[List[TradeRecord]] = None) -> Optional[TradeMetrics]:
        if trades is None:
            trades = self.load_trades()

        if not trades:
            return None

        total_trades = len(trades)
        wins = [t for t in trades if t.profit > 0]
        losses = [t for t in trades if t.profit < 0]

        winning_trades = len(wins)
        losing_trades = len(losses)
        win_rate = (winning_trades / total_trades) * 100.0 if total_trades > 0 else 0.0

        total_profit = sum(t.profit for t in trades)
        gross_profit = sum(t.profit for t in wins) if wins else 0.0
        gross_loss = abs(sum(t.profit for t in losses)) if losses else 0.0

        profit_factor = (gross_profit / gross_loss) if gross_loss > 0 else (999.0 if gross_profit > 0 else 0.0)

        # Max Drawdown 計算
        cumulative_profit = 0.0
        peak = 0.0
        max_drawdown = 0.0

        for t in trades:
            cumulative_profit += t.profit
            if cumulative_profit > peak:
                peak = cumulative_profit
            drawdown = peak - cumulative_profit
            if drawdown > max_drawdown:
                max_drawdown = drawdown

        max_drawdown_pct = (max_drawdown / (peak + 100000.0)) * 100.0 if peak > 0 else 0.0
        avg_trade_profit = (total_profit / total_trades) if total_trades > 0 else 0.0
        largest_win = max([t.profit for t in wins], default=0.0)
        largest_loss = min([t.profit for t in losses], default=0.0)

        return TradeMetrics(
            total_trades=total_trades,
            winning_trades=winning_trades,
            losing_trades=losing_trades,
            win_rate=win_rate,
            total_profit=total_profit,
            gross_profit=gross_profit,
            gross_loss=gross_loss,
            profit_factor=profit_factor,
            max_drawdown=max_drawdown,
            max_drawdown_pct=max_drawdown_pct,
            avg_trade_profit=avg_trade_profit,
            largest_win=largest_win,
            largest_loss=largest_loss,
            trades=trades,
        )

File: internal/usecase/analyzer.go
package usecase

import (
	"math"
	"rakutenfx/internal/domain"
)

type TradeAnalyzer struct{}

func NewTradeAnalyzer() *TradeAnalyzer {
	return &TradeAnalyzer{}
}

// CalculateMetrics calculates comprehensive performance metrics from closed trade records.
func (a *TradeAnalyzer) CalculateMetrics(trades []domain.TradeRecord) domain.TradeMetrics {
	if len(trades) == 0 {
		return a.GenerateBaselineMetrics()
	}

	var winningCount, losingCount int
	var grossProfit, grossLoss, totalProfit float64
	var largestWin, largestLoss float64
	var peakEquity, currentEquity, maxDrawdown float64
	var consecutiveWins, maxWins, consecutiveLosses, maxLosses int

	for _, t := range trades {
		p := t.Profit
		totalProfit += p
		currentEquity += p

		if currentEquity > peakEquity {
			peakEquity = currentEquity
		}
		dd := peakEquity - currentEquity
		if dd > maxDrawdown {
			maxDrawdown = dd
		}

		if p > 0 {
			winningCount++
			grossProfit += p
			if p > largestWin {
				largestWin = p
			}
			consecutiveWins++
			consecutiveLosses = 0
			if consecutiveWins > maxWins {
				maxWins = consecutiveWins
			}
		} else if p < 0 {
			losingCount++
			grossLoss += math.Abs(p)
			if p < largestLoss {
				largestLoss = p
			}
			consecutiveLosses++
			consecutiveWins = 0
			if consecutiveLosses > maxLosses {
				maxLosses = consecutiveLosses
			}
		}
	}

	totalTrades := len(trades)
	winRate := 0.0
	if totalTrades > 0 {
		winRate = (float64(winningCount) / float64(totalTrades)) * 100.0
	}

	pf := 0.0
	if grossLoss > 0 {
		pf = grossProfit / grossLoss
	} else if grossProfit > 0 {
		pf = 99.9
	}

	avgTradeProfit := 0.0
	if totalTrades > 0 {
		avgTradeProfit = totalProfit / float64(totalTrades)
	}

	maxDDPct := 0.0
	if peakEquity > 0 {
		maxDDPct = (maxDrawdown / peakEquity) * 100.0
	}

	// 推奨ロット動的計算 (0.10 ~ 0.50 Lot)
	recommendedLot := a.CalculateRecommendedLot(winRate, pf, maxLosses)

	return domain.TradeMetrics{
		TotalTrades:       totalTrades,
		WinningTrades:     winningCount,
		LosingTrades:      losingCount,
		WinRate:           math.Round(winRate*10) / 10,
		TotalProfit:       math.Round(totalProfit*100) / 100,
		GrossProfit:       math.Round(grossProfit*100) / 100,
		GrossLoss:         math.Round(grossLoss*100) / 100,
		ProfitFactor:      math.Round(pf*100) / 100,
		MaxDrawdown:       math.Round(maxDrawdown*100) / 100,
		MaxDrawdownPct:    math.Round(maxDDPct*10) / 10,
		AvgTradeProfit:    math.Round(avgTradeProfit*10) / 10,
		LargestWin:        largestWin,
		LargestLoss:       largestLoss,
		RecommendedLot:    recommendedLot,
		ConsecutiveWins:   maxWins,
		ConsecutiveLosses: maxLosses,
		Trades:            trades,
	}
}

// CalculateRecommendedLot determines dynamic position sizing based on recent edge.
func (a *TradeAnalyzer) CalculateRecommendedLot(winRate, pf float64, consecutiveLosses int) float64 {
	baseLot := 0.25

	// 連敗が多い場合は安全マージンを確保
	if consecutiveLosses >= 3 {
		return 0.10
	}

	// 高勝率かつ高PFの場合はロットを段階的に引き上げ
	if winRate >= 70.0 && pf >= 2.5 {
		return 0.50
	} else if winRate >= 60.0 && pf >= 1.8 {
		return 0.35
	} else if winRate < 45.0 || pf < 1.0 {
		return 0.10
	}

	return baseLot
}

// GenerateBaselineMetrics returns demo data when database has no closed trades.
func (a *TradeAnalyzer) GenerateBaselineMetrics() domain.TradeMetrics {
	p1, p2, p3 := 158.80, 158.94, 159.05
	t1, t2, t3 := "2026-08-20 10:30:00", "2026-08-20 11:15:00", "2026-08-20 12:45:00"

	demoTrades := []domain.TradeRecord{
		{
			Ticket:     10001,
			Symbol:     "USDJPY",
			Action:     "BUY",
			Lots:       0.5,
			OpenPrice:  158.50,
			ClosePrice: &p1,
			OpenTime:   "2026-08-20 10:00:00",
			CloseTime:  &t1,
			Profit:     15000.0,
			Comment:    "AutoOrder_FibDow_618",
			CreatedAt:  "2026-08-20 10:30:00",
		},
		{
			Ticket:     10002,
			Symbol:     "USDJPY",
			Action:     "SELL",
			Lots:       0.5,
			OpenPrice:  158.90,
			ClosePrice: &p2,
			OpenTime:   "2026-08-20 11:00:00",
			CloseTime:  &t2,
			Profit:     -2000.0,
			Comment:    "AutoOrder_FibDow_Cut4pips",
			CreatedAt:  "2026-08-20 11:15:00",
		},
		{
			Ticket:     10003,
			Symbol:     "USDJPY",
			Action:     "BUY",
			Lots:       0.5,
			OpenPrice:  158.55,
			ClosePrice: &p3,
			OpenTime:   "2026-08-20 12:00:00",
			CloseTime:  &t3,
			Profit:     25000.0,
			Comment:    "AutoOrder_FibDow_FE1618",
			CreatedAt:  "2026-08-20 12:45:00",
		},
	}

	return a.CalculateMetrics(demoTrades)
}
